### PR TITLE
DATA: Repository

### LINKED ISSUE #4: DATA: Repository
## Model:  Repository

| Field            | Type                   | Description                                                 |
| ---------------- | ---------------------- | ----------------------------------------------------------- |
| Id               | Guid                   | Our internal identifier                                      |
| Name             | string (Required, max 255) | The repository name                                           |
| Owner            | string (Required, max 255) | The name that represents the user or organization            |
| ExternalId       | string (Required, max 255) | The external unique identifier                                |
| SourceId           | Guid   | The Id of the source                 |
| IsOrganization   | bool                   | A flag that indicates if this is a user or organization       |
| IsPrivate        | bool                   | A flag that indicates if this is a private repository         |
| Token            | string                 | The token that should be used to access private repos         |
| TokenExpireAt      | DateTimeOffset         | When the token will expire                                    |
| Description     | string         | A description of the repo                                    |
| CreatedAt     | DateTimeOffset         | When the remote repo was created     |
| UpdatedAt     | DateTimeOffset         | When the remote repo was last updated    |

** **_Extra considerations:  Add a unique constraint on (ExternalId, Source).  This will ensure we do not have duplicate repositories._**

image

### PR DESCRIPTION
**Pull request by AmineTissilguit**
### Migration Generation Behavior
When I used the command `Add-Migration...` to generate migrations, I noticed that it automatically applied the migrations. This behavior is unexpected for me because my usual workflow involves generating migrations first to review the changes. I then apply the migrations only if everything looks correct. If not, I remove them using `Remove-Migration` without applying them. However, in this case, the `Add-Migration` command generated and applied the migration, resulting in the creation of two migration files. Could you please clarify if this is the expected behavior or if there might be a configuration issue on my end?

### Best Practices
What is your preferred naming convention for migrations?

Closes #4

### CODE REVIEW

**Comment by Lezune** [comment]
Hello @AmineTissilguit! Unfortunately I am unsure of the answer to your question, for me in particular, I followed the youtube video that was released yesterday. There is a document regarding best practices for "The-Standard-Organization" projects which you might like to read. I will leave the link here --> [The-Standard-Team](https://github.com/hassanhabib/The-Standard-Team)

File: GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿using GitFyle.Core.Api.Models.Foundations.Sources;
```
Comment (hassanhabib) [inline GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs:1]: Add copyrights

File: GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+using System;
```
Comment (hassanhabib) [inline GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs:2]: `System` goes on top always

File: GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace GitFyle.Core.Api.Models.Foundations.Repositories;
+
+public class Repository
+{
+    public Guid Id { get; set; }
+
+    [Required]
+    [MaxLength(255)]
+    public string Name { get; set; }
```
Comment (hassanhabib) [inline GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs:13]: Let's remove validations at the model level they should be here. As annotations. Let's handle this in Service/Foundation layer and the fluent rule enforcement at the configuration level.

File: GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    [MaxLength(255)]
+    public string ExternalId { get; set; }
+    public Guid SourceId { get; set; }
+    public bool IsOrganization { get; set; }
+    public bool IsPrivate { get; set; }
+    public string Token { get; set; }
+    public DateTimeOffset TokenExpireAt { get; set; }
+    public string Description { get; set; }
+
+    //Navigations
```
Comment (hassanhabib) [inline GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs:29]: Nooo comments

File: GitFyle.Core.Api/Models/Foundations/Sources/Source.cs [hunk 5]
```diff
@@ -11,5 +12,6 @@ public class Source
         public Guid Id { get; set; }
         public string Name { get; set; }
         public string Url { get; set; }
+        public Repository Repository { get; set; }
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Models/Foundations/Sources/Source.cs:15]: A source (GitHub or ...) can have many repositories, so we need to change this to:

```suggestion
        public IEnumerable<Repository> Repositories { get; set; }
```

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.Configurations.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            .HasMaxLength(255)
+            .IsRequired();
+
+        builder.HasIndex(contribution => new
+        {
+            contribution.Name,
+            contribution.Owner,
+            contribution.ExternalId,
+            contribution.SourceId
+        });
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.Configurations.cs:32]:
```suggestion
        builder.HasIndex(repository => new
        {
            repository.Name,
            repository.Owner,
            repository.ExternalId,
            repository.SourceId
        })
        .IsUnique();
```

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.Configurations.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+        builder.HasOne(x => x.Source)
+            .WithOne(x => x.Repository);
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.Configurations.cs:35]: This needs to change to a one-to-many relationship

```suggestion
            builder.HasOne(repository => repository.Source)
                .WithMany(source => source.Repositories)
                .HasForeignKey(repository => repository.SourceId)
                .OnDelete(DeleteBehavior.NoAction);
```

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
+﻿// ----------------------------------------------------------------------------------
+// Copyright (c) The Standard Organization: A coalition of the Good-Hearted Engineers
+// ----------------------------------------------------------------------------------
+
+using GitFyle.Core.Api.Models.Foundations.Repositories;
+using Microsoft.EntityFrameworkCore;
+
+namespace GitFyle.Core.Api.Brokers.Storages;
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.cs:8]:
```suggestion
namespace GitFyle.Core.Api.Brokers.Storages
{
```
and add the closing tag at the bottom.

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.Configurations.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
+﻿// ----------------------------------------------------------------------------------
+// Copyright (c) The Standard Organization: A coalition of the Good-Hearted Engineers
+// ----------------------------------------------------------------------------------
+
+using GitFyle.Core.Api.Models.Foundations.Repositories;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+
+namespace GitFyle.Core.Api.Brokers.Storages;
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.Configurations.cs:8]:
```suggestion
namespace GitFyle.Core.Api.Brokers.Storages
{
```
and add the closing tag at the bottom.

File: GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs [hunk 10]
```diff
@@ -0,0 +1,24 @@
+﻿// ----------------------------------------------------------------------------------
+// Copyright (c) The Standard Organization: A coalition of the Good-Hearted Engineers
+// ----------------------------------------------------------------------------------
+
+using System;
+using GitFyle.Core.Api.Models.Foundations.Sources;
+
+namespace GitFyle.Core.Api.Models.Foundations.Repositories;
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs:8]:
```suggestion
namespace GitFyle.Core.Api.Models.Foundations.Repositories;
{
```
and add the closing tag at the bottom.

File: GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+        public bool IsOrganization { get; set; }
+        public bool IsPrivate { get; set; }
+        public string Token { get; set; }
+        public DateTimeOffset TokenExpireAt { get; set; }
+        public string Description { get; set; }
+        public Source Source { get; set; }
+        public DateTimeOffset CreatedAt { get; set; }
+        public DateTimeOffset UpdatedAt { get; set; }
+    }
+
```
Comment (glhays) [inline GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs:26]: Remove extra line

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.Configurations.cs [hunk 12]
```diff
@@ -0,0 +1,42 @@
+﻿// ----------------------------------------------------------------------------------
+// Copyright (c) The Standard Organization: A coalition of the Good-Hearted Engineers
+// ----------------------------------------------------------------------------------
+
+using GitFyle.Core.Api.Models.Foundations.Repositories;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.Configurations.cs:6]:
```suggestion
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;
```

File: GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.Configurations.cs [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+                repository.Owner,
+                repository.ExternalId,
+                repository.SourceId
+            });
+
+            builder.HasOne(x => x.Source)
+                .WithMany(x => x.Repositories)
+                .HasForeignKey(x => x.SourceId)
+                .OnDelete(Microsoft.EntityFrameworkCore.DeleteBehavior.NoAction);
+        }
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Brokers/Storages/StorageBroker.Repositories.Configurations.cs:38]:
```suggestion
            builder.HasOne(x => x.Source)
                .WithMany(x => x.Repositories)
                .HasForeignKey(x => x.SourceId)
                .OnDelete(Microsoft.EntityFrameworkCore.DeleteBehavior.NoAction);
```

File: GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+        public string Name { get; set; }
+        public string Owner { get; set; }
+        public string ExternalId { get; set; }
+        public Guid SourceId { get; set; }
+        public bool IsOrganization { get; set; }
+        public bool IsPrivate { get; set; }
+        public string Token { get; set; }
+        public DateTimeOffset TokenExpireAt { get; set; }
+        public string Description { get; set; }
+        public Source Source { get; set; }
```
Comment (cjdutoit) [inline GitFyle.Core.Api/Models/Foundations/Repositories/Repository.cs:22]: Please move the navigation property to the bottom of the class. Placing navigation properties at the end helps to clearly separate the scalar properties (like Id, Name, etc.) from the navigation properties. This makes the class easier to read and understand at a glance.

**Review by cjdutoit — APPROVED** [review]

**Pull request merged**